amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeServiceImpl.java

```diff
@@ -0,0 +1,208 @@
+            sb.append("Attempt to redeclare default exchange: of type ")
+                    .append(ExchangeDefaults.DIRECT_EXCHANGE_CLASS);
+            channel.closeChannel(ErrorCodes.ACCESS_REFUSED, sb.toString());
+        }
+    }
+
+    private String formatString(String s) {
+        return s.replaceAll("\r", "").
+                replaceAll("\n", "").trim();
+    }
+
+    @Override
+    public void exchangeDeclare(AmqpChannel channel, AMQShortString exchange, AMQShortString type,
+                                boolean passive, boolean durable, boolean autoDelete,
+                                boolean internal, boolean nowait, FieldTable arguments) {
+        int channelId = channel.getChannelId();
+        AmqpConnection connection = channel.getConnection();
+        if (log.isDebugEnabled()) {
+            log.debug("RECV[{}] ExchangeDeclare[ exchange: {},"
+                            + " type: {}, passive: {}, durable: {}, autoDelete: {}, internal: {}, "
+                            + "nowait: {}, arguments: {} ]", channelId, exchange,
+                    type, passive, durable, autoDelete, internal, nowait, arguments);
+        }
+
+        handleDefaultExchangeInExchangeDeclare(channel, exchange);
+
+        String exchangeName = formatString(exchange.toString());
+        final AMQMethodBody declareOkBody = connection.getMethodRegistry().createExchangeDeclareOkBody();
+        boolean createIfMissing = !passive;
+        String exchangeType = type.toString();
+        if (channel.isBuildInExchange(exchange)) {
+            createIfMissing = true;
+            exchangeType = channel.getExchangeType(exchange.toString());
+        }
+        CompletableFuture<AmqpExchange> amqpExchangeCompletableFuture =
+                exchangeContainer.asyncGetExchange(connection.getNamespaceName(), exchangeName,
+                        createIfMissing, exchangeType);
+        amqpExchangeCompletableFuture.whenComplete((amqpExchange, throwable) -> {
+            if (throwable != null) {
+                log.error("Get Topic error:{}", throwable.getMessage());
+                connection.sendConnectionClose(ErrorCodes.NOT_FOUND, "Unknown exchange: " + exchangeName, channelId);
+            } else {
+                if (null == amqpExchange) {
+                    channel.closeChannel(ErrorCodes.NOT_FOUND, "Unknown exchange:" + exchangeName);
+                } else {
+                    if (!(type == null || type.length() == 0)
+                            && !amqpExchange.getType().toString().equalsIgnoreCase(type.toString())) {
+                        connection.sendConnectionClose(ErrorCodes.NOT_ALLOWED,
+                                "Attempt to redeclare exchange: '"
+                                        + exchangeName + "' of type " + amqpExchange.getType()
+                                        + " to " + type + ".", channelId);
+                    } else if (!nowait) {
+                        channel.sync();
```

Comment:
We can't call a sync method in the async callback.

Reply:
OK. sync() maybe useless, I will remove it.